pkg/storj/metainfo.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package storj
+
+import (
+	"context"
+)
+
+// Metainfo represents a database for storing meta-info about objects
+type Metainfo interface {
+	// MetainfoLimits returns limits for this metainfo database
+	Limits() (MetainfoLimits, error)
+
+	// CreateBucket creates a new bucket with the specified information
+	// Database automatically sets different values in the information
+	CreateBucket(ctx context.Context, bucket string, info *Bucket) (Bucket, error)
+	// DeleteBucket deletes bucket
+	DeleteBucket(ctx context.Context, bucket string) error
+	// GetBucket gets bucket information
+	GetBucket(ctx context.Context, bucket string) (Bucket, error)
+
+	// GetObject returns information about an object
+	GetObject(ctx context.Context, bucket string, path Path) (Object, error)
+	// GetObjectStream returns interface for reading the object stream
+	GetObjectStream(ctx context.Context, bucket string, path Path) (ReadonlyStream, error)
+
+	// CreateObject creates an uploading object and returns an interface for uploading Object information
+	CreateObject(ctx context.Context, bucket string, path Path, info Object) (MutableObject, error)
+	// ModifyObject creates an interface for modifying an existing object
+	ModifyObject(ctx context.Context, bucket string, path Path, info Object) (MutableObject, error)
+	// DeleteObject deletes an object from database
+	DeleteObject(ctx context.Context, bucket string, path Path) error
+
+	// TODO: add things for continuing existing objects
+	ListObjects(cctx context.Context, bucket string, options ListOptions) (ObjectList, error)
+}
+
+// ListOptions lists objects
+type ListOptions struct {
+	Prefix    Path
+	First     Path // First is relative to Prefix, full path is Prefix + First
+	Delimiter rune
+	Recursive bool
+	Limit     int
+}
+
+// ObjectList is a list of objects
+type ObjectList struct {
+	Bucket string
+	Prefix Path
+	More   bool
+
+	// Items paths are relative to Prefix
+	// To get the full path use list.Prefix + list.Items[0].Path
+	Items []Object
+}
+
+// MetainfoLimits lists limits specified for the Metainfo database
+type MetainfoLimits struct {
+	// ListLimit specifies the maximum amount of items that can be listed at a time.
+	ListLimit int64
+
+	// MaximumInlineSegment specifies the maximum inline segment that is allowed to be stored.
+	MaximumInlineSegmentSize int64
+}
+
+// ReadonlyStream is an interface for reading segment information
+type ReadonlyStream interface {
```

Comment:
"read only" is not a single word, so this type name should be rather "ReadOnlyStream".